_transcripts/032-rust-1.48-1.49.md

```diff
@@ -0,0 +1,731 @@
+talk more about those?
+
+__Jon__: No, I think those are fairly straightforward.
+
+__Ben__: We'll link to them if you want to read about them.
+
+__Jon__: Yeah, there are new some new stable functions for `slice`, which are
+basically, like, implementation details for binary search. Like, if I remember
+correctly, these are— so, `select_nth_unstable`, for example, which is one of
+the ones that stabilized, is: reorder the slice such that the element at the
+index that you give is at its final sorted position. So these are highly
+specialized functions that—
+
+__Ben__: Oh, jeez.
+
+__Jon__: Yeah, it's like— I think it's basically for, if you want to implement
+quicksort, you need this function. I'm not entirely sure why it's in the
+standard library, but I guess it's useful for different implementations. Not
+entirely sure.
+
+There are, though, a couple of other interesting things in the changelog, that—
+
+__Ben__: Hidden away, tucked away.
+
+__Jon__: Sorry, say that again?
+
+__Ben__: They're hidden away, they're tucked away secretly behind the link into
+the detailed change notes.
+
+__Jon__: Yeah, and often things are hidden away for a good reason, like they're
+very niche and only people like me start digging into them. But this time
+there's actually one that's really cool, and that is, in 1.49 `move_ref_pattern`
+was stabilized, and that might not mean much to you, but I promise you it's
+cool. So you may sometimes have written, like, a `match` statement, like let's
+say you're matching on `foo`. And `foo` is a type that happens to have multiple
+fields. Let's say it has the fields `x` and `y`. Previously, you had the choice
+of either matching those fields by reference or matching them by move, right? So
+if you do a `match foo`, then you can write the pattern as `Foo {x, y}` and then
+in the match arm, you can either, sort of move `x` and `y` out of `foo` and into
+that scope. Or you can just move them by reference so that after the match exits
+you can continue to use `foo`. And that's how things were up to 1.49.
+
+But what you were unable to do before, was to say, I want to move `y` out of
+`foo`, but I want to take `x` by reference. And if you think about it, this is a
+little bit of a weird operation, because what state is `foo` at, during and
+after those scopes, right? Like, if you do `match foo` and then the actual
+pattern is `Foo {ref x, y}`, then inside of the scope, if you now try to refer
+to `foo`, will that work or not? Right? Because you've moved, `y` out of `foo`.
+So it's sort of a partial move. The `x` is still there, and you have a reference
+to it. But the `y` is not. And this is actually something that they wanted to
+support this sort of mixing of move and ref bindings, all the way before Rust
+1.0. But the borrow checker just wasn't up to the task of ensuring that you
+wouldn't try to then use `foo.y` indirectly or directly, later on in your code.
```

Comment:
I've interpreted this as though there is a local variable `foo` of type `Foo`. Does that seem right?

Reply:
Yes, exactly right.